scrapy/signalmanager.py

```diff
@@ -3,15 +3,36 @@
 
 from scrapy.utils import signal as _signal
 from scrapy.exceptions import ScrapyDeprecationWarning
-from scrapy.dispatch.robustapply import robustApply
+from scrapy.dispatch.robustapply import robust_apply as _robust_apply
 from scrapy.dispatch.utils import func_accepts_kwargs
-
+from scrapy.dispatch import Signal as _Signal
 
 class SignalManager(object):
 
     def __init__(self, sender=None):
         self.sender = sender
-        self.signals = {}
+        self._patched_receivers = {}
+        self._signal_proxies = {}
+
+    def _ensure_signal(self, signal):
+        """
+        This method is for backward compatability for any custom signals
+        that might be in use in third party extensions which is still using the
+        standard python objects as signals. This method returns a Signal()
+        class instance and saves it to proxy the user defined signal.
+        """
+        if isinstance(signal, _Signal):
+            return signal
+        # Ensure we got at least an old style signal
+        warnings.warn("Signals in scrapy are no longer instances of the basic "
+                      "python object but rather the Signal class defined in "
+                      "the scrapy.dispatch module. Please refer to the Signal "
+                      "API documentation.",
+                      ScrapyDeprecationWarning, stacklevel=3)
+        signal_name = signal.__repr__()
+        if signal_name not in self._signal_proxies:
+            self._signal_proxies.setdefault(signal_name, _Signal())
+        return self._signal_proxies[signal_name]
```

Comment:
Sorry, I only read your [earlier comment](https://github.com/scrapy/scrapy/pull/2030#discussion_r71973524) just now. I would still argue strongly for using the instance itself.

* In broader terms, for the wisdom mentioned in the top answer in [the SO thread you mention](https://stackoverflow.com/questions/11162201/is-it-always-faster-to-use-string-as-key-in-a-dict) ("premature optimization is the root of all evil", a trap I fall for on a regular basis ;)).
* In this case in particular, because calculating an objects representation is a very expensive operation. Its costs vastly exceed the savings from using a string, making the `repr`-based dictionary almost 20 times slower:
  ```
In [17]: timeit.timeit('mydict[a]', 'a = object(); mydict = {a: True}')
Out[17]: 0.06446212899936654

In [18]: timeit.timeit('mydict[repr(a)]', 'a = object(); mydict = {repr(a): True}')
Out[18]: 1.168589578999672
    ```